pkg/internal/ebpf/capabilitytracer/capabilitytracer.go

```diff
@@ -0,0 +1,239 @@
+	beyla_ebpf "github.com/grafana/beyla/pkg/internal/ebpf"
+	ebpfcommon "github.com/grafana/beyla/pkg/internal/ebpf/common"
+	"github.com/grafana/beyla/pkg/internal/exec"
+	"github.com/grafana/beyla/pkg/internal/goexec"
+	"github.com/grafana/beyla/pkg/internal/imetrics"
+	"github.com/grafana/beyla/pkg/internal/request"
+	"github.com/grafana/beyla/pkg/internal/svc"
+)
+
+//go:generate $BPF2GO -cc $BPF_CLANG -cflags $BPF_CFLAGS -type capability_info_t -target amd64,arm64 bpf ../../../../bpf/capability_tracer.c -- -I../../../../bpf/headers
+//go:generate $BPF2GO -cc $BPF_CLANG -cflags $BPF_CFLAGS -type capability_info_t -target amd64,arm64 bpf_tp ../../../../bpf/capability_tracer.c -- -I../../../../bpf/headers -DBPF_TRACEPARENT
+//go:generate $BPF2GO -cc $BPF_CLANG -cflags $BPF_CFLAGS -type capability_info_t -target amd64,arm64 bpf_debug ../../../../bpf/capability_tracer.c -- -I../../../../bpf/headers -DBPF_DEBUG
+//go:generate $BPF2GO -cc $BPF_CLANG -cflags $BPF_CFLAGS -type capability_info_t -target amd64,arm64 bpf_tp_debug ../../../../bpf/capability_tracer.c -- -I../../../../bpf/headers -DBPF_DEBUG -DBPF_TRACEPARENT
+
+type BPFCapabilityInfo bpfCapabilityInfoT
+
+type Tracer struct {
+	pidsFilter ebpfcommon.ServiceFilter
+	cfg        *beyla.Config
+	bpfObjects bpfObjects
+	closers    []io.Closer
+	log        *slog.Logger
+	metrics    imetrics.Reporter
+}
+
+// AddInstrumentedLibRef implements ebpf.Tracer.
+func (p *Tracer) AddInstrumentedLibRef(uint64) {
+}
+
+// Updating these requires updating the constants below in pid.h
+// #define MAX_CONCURRENT_PIDS 3001 // estimate: 1000 concurrent processes (including children) * 3 namespaces per pid
+// #define PRIME_HASH 192053 // closest prime to 3001 * 64
+const (
+	maxConcurrentPids = 3001
+	primeHash         = 192053
+)
+
+func pidSegmentBit(k uint64) (uint32, uint32) {
+	h := uint32(k % primeHash)
+	segment := h / 64
+	bit := h & 63
+
+	return segment, bit
+}
+
+func (p *Tracer) buildPidFilter() []uint64 {
+	result := make([]uint64, maxConcurrentPids)
+	for nsid, pids := range p.pidsFilter.CurrentPIDs(ebpfcommon.PIDTypeKProbes) {
+		for pid := range pids {
+			// skip any pids that might've been added, but are not tracked by the kprobes
+			p.log.Debug("Reallowing pid", "pid", pid, "namespace", nsid)
+
+			k := uint64((uint64(nsid) << 32) | uint64(pid))
+
+			segment, bit := pidSegmentBit(k)
+
+			v := result[segment]
+			v |= (1 << bit)
+			result[segment] = v
+		}
+	}
+
+	return result
+}
+
+func (p *Tracer) rebuildValidPids() {
+	if p.bpfObjects.ValidPids != nil {
+		v := p.buildPidFilter()
+
+		p.log.Debug("number of segments in pid filter cache", "len", len(v))
+
+		for i, segment := range v {
+			err := p.bpfObjects.ValidPids.Put(uint32(i), uint64(segment))
+			if err != nil {
+				p.log.Error("Error setting up pid in BPF space, sizes of Go and BPF maps don't match", "error", err, "i", i)
+			}
+		}
+	}
+}
+
+func (p *Tracer) AllowPID(pid, ns uint32, svc *svc.Attrs) {
```

Comment:
It takes a bit of time for this filtering to kick in. Therefore, I think Beyla might miss kprobes triggering right when an Alloy process starts up. Not sure what the best way to handle this is.